truffle/src/main/java/org/jruby/truffle/nodes/core/ObjectSpaceNodes.java

```diff
@@ -65,32 +62,26 @@ public long id2RefSmallInt(long id) {
             return ObjectIDOperations.toFixnum(id);
         }
 
+        @TruffleBoundary
         @Specialization(guards = "isBasicObjectID(id)")
-        public Object id2Ref(final VirtualFrame frame, final long id,
-                @Cached("createReadObjectIDNode()") final ReadHeadObjectFieldNode readObjectIdNode) {
-            CompilerDirectives.transferToInterpreter();
+        public DynamicObject id2Ref(
+                final long id,
+                @Cached("createReadObjectIDNode()") ReadHeadObjectFieldNode readObjectIdNode) {
+            for (DynamicObject object : ObjectGraph.stopAndGetAllObjects(this, getContext())) {
+                final long objectID;
+
+                try {
+                    objectID = readObjectIdNode.executeLong(object);
+                } catch (UnexpectedResultException e) {
+                    throw new UnsupportedOperationException(e);
+                }
 
-            final Memo<Object> result = new Memo<Object>(nil());
-
-            new ObjectGraph(getContext()).visitObjects(new ObjectGraphVisitor() {
-                @Override
-                public boolean visit(DynamicObject object) throws StopVisitingObjectsException {
-                    final long objectID;
-                    try {
-                        objectID = readObjectIdNode.executeLong(object);
-                    } catch (UnexpectedResultException e) {
-                        throw new UnsupportedOperationException(e);
-                    }
-
-                    if (objectID == id) {
-                        result.set(object);
-                        throw new StopVisitingObjectsException();
-                    }
-                    return true;
+                if (objectID == id) {
+                    return object;
                 }
-            });
+            }
 
-            return result.get();
+            return nil();
```

Comment:
I copied JRuby behaviour which returns `nil` - looks like a bug on their behalf. I'll add a spec for it.